### PR TITLE
[JBPM-9094] Integration tests for new customCaseIdPrefix ability

### PR DESCRIPTION
**Pull request by gmunozfe**

### CODE REVIEW

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-all/src/test/filtered-resources/kjars-sources/case-id-generator/src/main/resources/org/kie/server/testing/EmptyCaseIdExpressionPrefixUpper.bpmn2 [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+             xsi:schemaLocation="http://www.omg.org/spec/BPMN/20100524/MODEL BPMN20.xsd"
+             xmlns:bpmndi="http://www.omg.org/spec/BPMN/20100524/DI"
+             xmlns:dc="http://www.omg.org/spec/DD/20100524/DC"
+             xmlns:di="http://www.omg.org/spec/DD/20100524/DI"
+             xmlns:tns="http://www.jboss.org/drools">
+
+  <itemDefinition id="_sItem" structureRef="String" />
+  <itemDefinition id="_caseFile_CaseIdItem" structureRef="String"/>
+  <itemDefinition id="_caseFile_outcomeItem" structureRef="String"/>
+  <process processType="Private" isExecutable="true" id="EmptyCaseIdExpressionPrefixUpper" name="EmptyCaseIdExpressionPrefixUpper" tns:adHoc="true">
```
Comment (afalhambra) [inline kie-server-parent/kie-server-tests/kie-server-integ-tests-all/src/test/filtered-resources/kjars-sources/case-id-generator/src/main/resources/org/kie/server/testing/EmptyCaseIdExpressionPrefixUpper.bpmn2:17]:
```suggestion
  <process processType="Private" isExecutable="true" id="EmptyCaseWithUpperPrefixIdExpression" name="EmptyCaseWithUpperPrefixIdExpression" tns:adHoc="true">
```

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-all/src/test/filtered-resources/kjars-sources/case-id-generator/src/main/resources/org/kie/server/testing/EmptyCaseIdExpressionPrefixUpper.bpmn2 [hunk 2]
```diff
@@ -0,0 +1,32 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (afalhambra) [inline kie-server-parent/kie-server-tests/kie-server-integ-tests-all/src/test/filtered-resources/kjars-sources/case-id-generator/src/main/resources/org/kie/server/testing/EmptyCaseIdExpressionPrefixUpper.bpmn2:1]: would rename this to:

EmptyCaseWithUpperPrefixIdExpression.bpmn2

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-all/src/test/java/org/kie/server/integrationtests/jbpm/cases/CaseIdGeneratorIntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+    private static final String CASE_OWNER_ROLE = "owner";
+    
+    private static final String CORRELATION_KEY = "my correlation key";
+    
+    private static final String EMPTY_CASE_ID_CASE_PREFIX = "EmptyCaseWithIdCasePrefix";
+    private static final String EMPTY_CASE_ID_CASE_DEPLOYMENT = "EmptyCaseWithIdCaseDeployment";
+    private static final String EMPTY_CASE_ID_CASE_DEFINITION = "EmptyCaseWithIdCaseDefinition";
+    private static final String EMPTY_CASE_ID_VARIABLE_PREFIX = "EmptyCaseIdVariablePrefix";
+    private static final String EMPTY_CASE_ID_VARIABLE_PREFIX_NO_SEQ = "EmptyCaseIdVariablePrefixNoSequence";
+    private static final String EMPTY_CASE_ID_EXPRESSION_PREFIX_UPPER = "EmptyCaseIdExpressionPrefixUpper";
```
Comment (afalhambra) [inline kie-server-parent/kie-server-tests/kie-server-integ-tests-all/src/test/java/org/kie/server/integrationtests/jbpm/cases/CaseIdGeneratorIntegrationTest.java:51]:
```suggestion
    private static final String EMPTY_CASE_WITH_UPPER_PREFIX_ID_EXPRESSION = "EmptyCaseWithUpperPrefixIdExpression";
```

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-all/src/test/java/org/kie/server/integrationtests/jbpm/cases/CaseIdGeneratorIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+        data.put("CORRELATION_KEY", CORRELATION_KEY);
+        buildCaseFile();
+        
+        String caseId = caseClient.startCase(CONTAINER_ID, EMPTY_CASE_ID_VARIABLE_PREFIX_NO_SEQ, caseFile);
+        assertNotNull(caseId);
+        assertEquals(CORRELATION_KEY, caseId);
+    }
+    
+    @Test
+    public void testStartEmptyCaseImplicitVariableWithExpressionUpper() {
```
Comment (afalhambra) [inline kie-server-parent/kie-server-tests/kie-server-integ-tests-all/src/test/java/org/kie/server/integrationtests/jbpm/cases/CaseIdGeneratorIntegrationTest.java:160]:
```suggestion
    public void testStartEmptyCaseWithUpperPrefixIdExpression() {
```

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-all/src/test/filtered-resources/kjars-sources/case-id-generator/src/main/resources/org/kie/server/testing/EmptyCaseIdExpressionPrefixTruncate.bpmn2 [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+             xsi:schemaLocation="http://www.omg.org/spec/BPMN/20100524/MODEL BPMN20.xsd"
+             xmlns:bpmndi="http://www.omg.org/spec/BPMN/20100524/DI"
+             xmlns:dc="http://www.omg.org/spec/DD/20100524/DC"
+             xmlns:di="http://www.omg.org/spec/DD/20100524/DI"
+             xmlns:tns="http://www.jboss.org/drools">
+
+  <itemDefinition id="_sItem" structureRef="String" />
+  <itemDefinition id="_caseFile_CaseIdItem" structureRef="String"/>
+  <itemDefinition id="_caseFile_outcomeItem" structureRef="String"/>
+  <process processType="Private" isExecutable="true" id="EmptyCaseIdExpressionPrefixTruncate" name="EmptyCaseIdExpressionPrefixTruncate" tns:adHoc="true">
```
Comment (afalhambra) [inline kie-server-parent/kie-server-tests/kie-server-integ-tests-all/src/test/filtered-resources/kjars-sources/case-id-generator/src/main/resources/org/kie/server/testing/EmptyCaseIdExpressionPrefixTruncate.bpmn2:17]:
```suggestion
  <process processType="Private" isExecutable="true" id="EmptyCaseWithTruncatePrefixIdExpression" name="EmptyCaseWithTruncatePrefixIdExpression" tns:adHoc="true">
```

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-all/src/test/filtered-resources/kjars-sources/case-id-generator/src/main/resources/org/kie/server/testing/EmptyCaseIdExpressionPrefixTruncate.bpmn2 [hunk 6]
```diff
@@ -0,0 +1,32 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (afalhambra) [inline kie-server-parent/kie-server-tests/kie-server-integ-tests-all/src/test/filtered-resources/kjars-sources/case-id-generator/src/main/resources/org/kie/server/testing/EmptyCaseIdExpressionPrefixTruncate.bpmn2:1]: would rename to:

EmptyCaseWithTruncatePrefixIdExpression.bpmn2

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-all/src/test/java/org/kie/server/integrationtests/jbpm/cases/CaseIdGeneratorIntegrationTest.java [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+    
+    private static final String CORRELATION_KEY = "my correlation key";
+    
+    private static final String EMPTY_CASE_ID_CASE_PREFIX = "EmptyCaseWithIdCasePrefix";
+    private static final String EMPTY_CASE_ID_CASE_DEPLOYMENT = "EmptyCaseWithIdCaseDeployment";
+    private static final String EMPTY_CASE_ID_CASE_DEFINITION = "EmptyCaseWithIdCaseDefinition";
+    private static final String EMPTY_CASE_ID_VARIABLE_PREFIX = "EmptyCaseIdVariablePrefix";
+    private static final String EMPTY_CASE_ID_VARIABLE_PREFIX_NO_SEQ = "EmptyCaseIdVariablePrefixNoSequence";
+    private static final String EMPTY_CASE_ID_EXPRESSION_PREFIX_UPPER = "EmptyCaseIdExpressionPrefixUpper";
+    private static final String EMPTY_CASE_ID_EXPRESSION_PREFIX_TRUNCATE = "EmptyCaseIdExpressionPrefixTruncate";
```
Comment (afalhambra) [inline kie-server-parent/kie-server-tests/kie-server-integ-tests-all/src/test/java/org/kie/server/integrationtests/jbpm/cases/CaseIdGeneratorIntegrationTest.java:52]:
```suggestion
    private static final String EMPTY_CASE_WITH_TRUNCATE_PREFIX_ID_EXPRESSION = "EmptyCaseWithTruncatePrefixIdExpression";
```

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-all/src/test/java/org/kie/server/integrationtests/jbpm/cases/CaseIdGeneratorIntegrationTest.java [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+        data.put("type", "type1");
+        buildCaseFile();
+        String caseId = caseClient.startCase(CONTAINER_ID, EMPTY_CASE_ID_EXPRESSION_PREFIX_UPPER, caseFile);
+        assertNotNull(caseId);
+        assertTrue("Created Case Id doesn't match expected value, value is: " + caseId, 
+                    caseId.matches(EMPTY_CASE_ID_EXPRESSION_PREFIX_UPPER+"-TYPE1-[0-9]+"));
+    }
+    
+    @Test
+    public void testStartEmptyCaseImplicitVariableWithExpressionTruncate() {
```
Comment (afalhambra) [inline kie-server-parent/kie-server-tests/kie-server-integ-tests-all/src/test/java/org/kie/server/integrationtests/jbpm/cases/CaseIdGeneratorIntegrationTest.java:170]:
```suggestion
    public void testStartEmptyCaseWithTruncatePrefixIdExpression() {
```

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-all/src/test/filtered-resources/kjars-sources/case-id-generator/src/main/resources/org/kie/server/testing/EmptyCaseIdExpressionPrefixLpad.bpmn2 [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+             xsi:schemaLocation="http://www.omg.org/spec/BPMN/20100524/MODEL BPMN20.xsd"
+             xmlns:bpmndi="http://www.omg.org/spec/BPMN/20100524/DI"
+             xmlns:dc="http://www.omg.org/spec/DD/20100524/DC"
+             xmlns:di="http://www.omg.org/spec/DD/20100524/DI"
+             xmlns:tns="http://www.jboss.org/drools">
+
+  <itemDefinition id="_sItem" structureRef="String" />
+  <itemDefinition id="_caseFile_CaseIdItem" structureRef="String"/>
+  <itemDefinition id="_caseFile_outcomeItem" structureRef="String"/>
+  <process processType="Private" isExecutable="true" id="EmptyCaseIdExpressionPrefixLpad" name="EmptyCaseIdExpressionPrefixLpad" tns:adHoc="true">
```
Comment (afalhambra) [inline kie-server-parent/kie-server-tests/kie-server-integ-tests-all/src/test/filtered-resources/kjars-sources/case-id-generator/src/main/resources/org/kie/server/testing/EmptyCaseIdExpressionPrefixLpad.bpmn2:17]:
```suggestion
  <process processType="Private" isExecutable="true" id="EmptyCaseWithLpadPrefixIdExpression" name="EmptyCaseWithLpadPrefixIdExpression" tns:adHoc="true">
```

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-all/src/test/filtered-resources/kjars-sources/case-id-generator/src/main/resources/org/kie/server/testing/EmptyCaseIdExpressionPrefixLpad.bpmn2 [hunk 10]
```diff
@@ -0,0 +1,32 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (afalhambra) [inline kie-server-parent/kie-server-tests/kie-server-integ-tests-all/src/test/filtered-resources/kjars-sources/case-id-generator/src/main/resources/org/kie/server/testing/EmptyCaseIdExpressionPrefixLpad.bpmn2:1]: would rename to:

EmptyCaseWithLpadPrefixIdExpression.bpmn2

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-all/src/test/java/org/kie/server/integrationtests/jbpm/cases/CaseIdGeneratorIntegrationTest.java [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+    private static final String CORRELATION_KEY = "my correlation key";
+    
+    private static final String EMPTY_CASE_ID_CASE_PREFIX = "EmptyCaseWithIdCasePrefix";
+    private static final String EMPTY_CASE_ID_CASE_DEPLOYMENT = "EmptyCaseWithIdCaseDeployment";
+    private static final String EMPTY_CASE_ID_CASE_DEFINITION = "EmptyCaseWithIdCaseDefinition";
+    private static final String EMPTY_CASE_ID_VARIABLE_PREFIX = "EmptyCaseIdVariablePrefix";
+    private static final String EMPTY_CASE_ID_VARIABLE_PREFIX_NO_SEQ = "EmptyCaseIdVariablePrefixNoSequence";
+    private static final String EMPTY_CASE_ID_EXPRESSION_PREFIX_UPPER = "EmptyCaseIdExpressionPrefixUpper";
+    private static final String EMPTY_CASE_ID_EXPRESSION_PREFIX_TRUNCATE = "EmptyCaseIdExpressionPrefixTruncate";
+    private static final String EMPTY_CASE_ID_EXPRESSION_PREFIX_LPAD = "EmptyCaseIdExpressionPrefixLpad";
```
Comment (afalhambra) [inline kie-server-parent/kie-server-tests/kie-server-integ-tests-all/src/test/java/org/kie/server/integrationtests/jbpm/cases/CaseIdGeneratorIntegrationTest.java:53]:
```suggestion
    private static final String EMPTY_CASE_WITH_LPAD_PREFIX_ID_EXPRESSION = "EmptyCaseWithLpadPrefixIdExpression";
```

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-all/src/test/java/org/kie/server/integrationtests/jbpm/cases/CaseIdGeneratorIntegrationTest.java [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+        data.put("type", "type_very_very_long");
+        buildCaseFile();
+        String caseId = caseClient.startCase(CONTAINER_ID, EMPTY_CASE_ID_EXPRESSION_PREFIX_TRUNCATE, caseFile);
+        assertNotNull(caseId);
+        assertTrue("Created Case Id doesn't match expected value, value is: " + caseId, 
+                    caseId.matches(EMPTY_CASE_ID_EXPRESSION_PREFIX_TRUNCATE+"-type_very-[0-9]+"));
+    }
+
+    @Test
+    public void testStartEmptyCaseImplicitVariableWithExpressionLpad() {
```
Comment (afalhambra) [inline kie-server-parent/kie-server-tests/kie-server-integ-tests-all/src/test/java/org/kie/server/integrationtests/jbpm/cases/CaseIdGeneratorIntegrationTest.java:180]:
```suggestion
    public void testStartEmptyCaseWithLpadPrefixIdExpression() {
```

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-all/src/test/filtered-resources/kjars-sources/case-id-generator/src/main/resources/org/kie/server/testing/EmptyCaseIdExpressionPrefixRpad.bpmn2 [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+             xsi:schemaLocation="http://www.omg.org/spec/BPMN/20100524/MODEL BPMN20.xsd"
+             xmlns:bpmndi="http://www.omg.org/spec/BPMN/20100524/DI"
+             xmlns:dc="http://www.omg.org/spec/DD/20100524/DC"
+             xmlns:di="http://www.omg.org/spec/DD/20100524/DI"
+             xmlns:tns="http://www.jboss.org/drools">
+
+  <itemDefinition id="_sItem" structureRef="String" />
+  <itemDefinition id="_caseFile_CaseIdItem" structureRef="String"/>
+  <itemDefinition id="_caseFile_outcomeItem" structureRef="String"/>
+  <process processType="Private" isExecutable="true" id="EmptyCaseIdExpressionPrefixRpad" name="EmptyCaseIdExpressionPrefixRpad" tns:adHoc="true">
```
Comment (afalhambra) [inline kie-server-parent/kie-server-tests/kie-server-integ-tests-all/src/test/filtered-resources/kjars-sources/case-id-generator/src/main/resources/org/kie/server/testing/EmptyCaseIdExpressionPrefixRpad.bpmn2:17]:
```suggestion
  <process processType="Private" isExecutable="true" id="EmptyCaseWithRpadPrefixIdExpression" name="EmptyCaseWithRpadPrefixIdExpression" tns:adHoc="true">
```

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-all/src/test/filtered-resources/kjars-sources/case-id-generator/src/main/resources/org/kie/server/testing/EmptyCaseIdExpressionPrefixRpad.bpmn2 [hunk 14]
```diff
@@ -0,0 +1,32 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (afalhambra) [inline kie-server-parent/kie-server-tests/kie-server-integ-tests-all/src/test/filtered-resources/kjars-sources/case-id-generator/src/main/resources/org/kie/server/testing/EmptyCaseIdExpressionPrefixRpad.bpmn2:1]: would rename to:

EmptyCaseWithRpadPrefixIdExpression.bpmn2

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-all/src/test/java/org/kie/server/integrationtests/jbpm/cases/CaseIdGeneratorIntegrationTest.java [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+    
+    private static final String EMPTY_CASE_ID_CASE_PREFIX = "EmptyCaseWithIdCasePrefix";
+    private static final String EMPTY_CASE_ID_CASE_DEPLOYMENT = "EmptyCaseWithIdCaseDeployment";
+    private static final String EMPTY_CASE_ID_CASE_DEFINITION = "EmptyCaseWithIdCaseDefinition";
+    private static final String EMPTY_CASE_ID_VARIABLE_PREFIX = "EmptyCaseIdVariablePrefix";
+    private static final String EMPTY_CASE_ID_VARIABLE_PREFIX_NO_SEQ = "EmptyCaseIdVariablePrefixNoSequence";
+    private static final String EMPTY_CASE_ID_EXPRESSION_PREFIX_UPPER = "EmptyCaseIdExpressionPrefixUpper";
+    private static final String EMPTY_CASE_ID_EXPRESSION_PREFIX_TRUNCATE = "EmptyCaseIdExpressionPrefixTruncate";
+    private static final String EMPTY_CASE_ID_EXPRESSION_PREFIX_LPAD = "EmptyCaseIdExpressionPrefixLpad";
+    private static final String EMPTY_CASE_ID_EXPRESSION_PREFIX_RPAD = "EmptyCaseIdExpressionPrefixRpad";
```
Comment (afalhambra) [inline kie-server-parent/kie-server-tests/kie-server-integ-tests-all/src/test/java/org/kie/server/integrationtests/jbpm/cases/CaseIdGeneratorIntegrationTest.java:54]:
```suggestion
    private static final String EMPTY_CASE_WITH_RPAD_PREFIX_ID_EXPRESSION = "EmptyCaseWithRpadPrefixIdExpression";
```

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-all/src/test/java/org/kie/server/integrationtests/jbpm/cases/CaseIdGeneratorIntegrationTest.java [hunk 16]
```diff
@@ -0,0 +1,205 @@
...
+
+public class CaseIdGeneratorIntegrationTest extends JbpmKieServerBaseIntegrationTest {
+
+    private static ReleaseId releaseId = new ReleaseId("org.kie.server.testing", "case-id-generator", "1.0.0.Final");
+
+    private static final String CONTAINER_ID = "id-generator";
+
+    private static final String CASE_OWNER_ROLE = "owner";
+    
+    private static final String CORRELATION_KEY = "my correlation key";
```
[the remark this answers is not in the document]
Comment (gmunozfe) [reply kie-server-parent/kie-server-tests/kie-server-integ-tests-all/src/test/java/org/kie/server/integrationtests/jbpm/cases/CaseIdGeneratorIntegrationTest.java:44]: Yes, it's a free string

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-case-id-generator/src/main/java/org/kie/server/generator/CustomNumberPrefixCaseIdGenerator.java [hunk 17]
```diff
@@ -41,4 +60,13 @@ public void register(String prefix) {
     public void unregister(String prefix) {
         caseIdMap.remove(prefix);
     }
+    
+    @Override
+    public String resolveCaseIdPrefix(String expression, Map<String, Object> optionalParameters) {
+        try {
+            return !expression.isEmpty() ? (String) TemplateRuntime.eval(expression, CASE_ID_FUNCTIONS, optionalParameters) : "";
+        } catch (org.mvel2.CompileException e) {
```
[the remark this answers is not in the document]
Comment (gmunozfe) [reply kie-server-parent/kie-server-tests/kie-server-integ-tests-case-id-generator/src/main/java/org/kie/server/generator/CustomNumberPrefixCaseIdGenerator.java]: Done

**Comment by gmunozfe** [comment]
Thanks for your review @afalhambra : all comments have been resolved

**Comment by gmunozfe** [comment]
jenkins retest this

**Comment by gmunozfe** [comment]
jenkins retest this

**Comment by gmunozfe** [comment]
jenkins retest this please

**Comment by gmunozfe** [comment]
@kiegroup/gatekeepers could you please merge this?

**Pull request merged**